### DISCUSSION TITLE
Generator_loss  curve and discriminator_loss  curve looks wrong？

### DISCUSSION THREAD

**Discussion by jdwang125** [jhauret/eben#1] [Q&A] [answered]
On the basis of github code and the dataset of mls_french , add tenserboard to monitor the generator_loss curve and discriminator_loss  curve, maxepoch=50. From the result, the two loss curves are incorrect. The dataset is mls_french.
![image](https://user-images.githubusercontent.com/10845446/225520841-1753def9-f1f2-45eb-97fc-5510244e3c50.png)
![image](https://user-images.githubusercontent.com/10845446/225520866-7d4c1a57-7dfb-4a6a-aef5-df420163b92f.png)
![image](https://user-images.githubusercontent.com/10845446/225520895-e9206f9e-3dae-4195-ae9a-51b0e1e027c8.png)
![image](https://user-images.githubusercontent.com/10845446/225520918-3631cbcd-e5da-4906-8524-020c4db210dc.png)
![image](https://user-images.githubusercontent.com/10845446/225520935-2c209cf0-1103-4f92-86d6-7b28fca0cff3.png)
![image](https://user-images.githubusercontent.com/10845446/225520973-8777062a-d5ee-4219-a79a-d512214f3f56.png)
![image](https://user-images.githubusercontent.com/10845446/225521000-986becda-7876-4e9b-983c-2c5809d1d64b.png)

**Comment by jhauret** [comment, marked as answer]
Hi, 

Your curves are not incorrect, you just have pushed the training too far! :smile:  If you stop the training at the 13th epoch($\approx 200k$ steps), you will obtain the same model as the one in the project.

The real-time and on-device constraints have forced us to have a much lower number of generator parameters in comparison to discriminators (1.9M vs 27.8M). Starting from that point, it was hard to reach a Nash equilibrium during the training. BUT, this does not prevent obtaining a performing generator.

If you want to go further than a simple result reproduction, we have tried two interesting techniques from the [Encodec Paper](https://arxiv.org/abs/2210.13438) that helped to stabilize training and improved results:

-  **Normalizing the feature matching loss** with the reference embedding norm.
-  **Adding a dynamic loss balancer** that ensures generator parameters updates are equally impacted by reconstructive and adversarial losses (based on an exponential moving average normalization of gradients )

We did not include those techniques as they were not part of our original paper.

**Reply by jdwang125** [reply]
Thank you for your reply. Regarding the loss curve, I have two more questions:

1. Is it determined whether the model is well-trained based on `dev_D_adv_loss_valid = 0.5` (≈200k steps)?
2. Why do `dev_G_adv_loss` and `dev_G_ftr_loss` keep increasing instead of decreasing? Are they supposed to keep increasing?

Also, I have read the EnCodec paper. Could you please share the relevant training code with me via email at jdwang125@gmail.com? I would like to try it out. Thank you.

**Reply by jhauret** [reply]
Happy to help !

1. When using a GAN paradigm, you can't really tell when your generator is well-trained based on loss curves. 200k steps just correspond to two days of computing on a single Nvidia RTX 2080 Ti GPU which was the computational budget we used to compare the different approaches in the paper. This might not be the optimal number of steps to train for. 
To assess model performances you can use the STOI metric which on this very specific problem has shown to be OK... still, the best way to evaluate the audio quality/intelligibility is human ears :ear: 
2. Even though the generator is trained to minimize the feature matching loss, discriminators are also trained at the same time so their embeddings become more and more precise to distinguish real from enhanced audio, which causes this augmentation. Same for the adversarial part: discriminators are harder to confuse as training progress.


For your last request, here are the two requested functionalities taken from my development code:
````python
# the dynamic loss balancer
  def compute_ema_lambda_adaptive(self, loss_recons, loss_adv):
    """Compute a dynamic loss ponderation to minimize lambda*loss_recons+loss_adv
    Args:
        loss_recons: 1-element tensor corresponding to the feature matching loss
        loss_adv: 1-element tensor corresponding to the adversarial loss
    Return:
       lambda_adaptive_past: the adaptive ponderation coefficient
    """

      beta = 0.99

      # compute lambda_adaptive_t
      grads_recons = torch.autograd.grad(outputs=loss_recons, inputs=self.generator.last_conv.weight, retain_graph=True)[0]
      grads_adv = torch.autograd.grad(outputs=loss_adv, inputs=self.generator.last_conv.weight, retain_graph=True)[0]
      lambda_adaptive_new = torch.norm(grads_adv) / (torch.norm(grads_recons) + 1e-4)

      if self.lambda_adaptive_past is None:
          self.lambda_adaptive_past = lambda_adaptive_new
      else:  # Exponential Moving average
          self.lambda_adaptive_past = beta * self.lambda_adaptive_past + (1-beta)*lambda_adaptive_new

      self.lambda_adaptive_past = torch.clamp(self.lambda_adaptive_past, 0.0, 1e4).detach()

      return self.lambda_adaptive_past
````

````python         
 # the normalized feature matching loss
            ftr_loss = 0
            for scale in range(len(reference_embeddings)):  # across scales
                for layer in range(1, len(reference_embeddings[scale]) - 1):  # across layers
                    a = reference_embeddings[scale][layer]
                    b = enhanced_embeddings[scale][layer]
                    # ftr_loss += self.l1(a, b) / (len(reference_embeddings[scale]) - 2) #not normalized
                    ftr_loss += self.l1(a, b) / (torch.mean(torch.abs(a))*(len(target_embeddings[scale]) - 2)) # normalized

            ftr_loss /= len(reference_embeddings)`
````

**Reply by jdwang125** [reply]
Hi jhauret：
      After applying these two changes, the generator loss curve tends to stabilize, and it appears that convergence has been achieved this time.
![image](https://user-images.githubusercontent.com/10845446/228515105-4d8e8ab0-5589-496a-8d84-d669b9918f00.png)
![image](https://user-images.githubusercontent.com/10845446/228515178-e0f93634-7b20-4182-9dbd-a9d7e1677ef5.png)
![image](https://user-images.githubusercontent.com/10845446/228515245-0ba4b917-8610-4fdf-a4fb-33c491107f65.png)
![image](https://user-images.githubusercontent.com/10845446/228515286-1d461570-710a-4c4e-828a-94598b7763d1.png)
![image](https://user-images.githubusercontent.com/10845446/228515347-3f6df98d-b4f4-4d7e-9012-48111dcabded.png)
![image](https://user-images.githubusercontent.com/10845446/228515389-07e0fe14-c30a-42fb-bd73-9be06a9321a7.png)
![image](https://user-images.githubusercontent.com/10845446/228515492-e365934e-ece8-4b98-966b-aa47e316c328.png)

**Reply by andyye1999** [reply]
> ```python
> target_embeddings
> ```

what is target_embeddings mean？

**Reply by jhauret** [reply]
Good catch: it's a typo. It should be `reference_embeddings`.

**Reply by andyye1999** [reply]
in def compute_ema_lambda_adaptive  Is self.lambda_adaptive_past used to replace the 100 in gen_loss = adv_loss + 100 * ftr_loss?

**Reply by jhauret** [reply]
Yes it is !

**Reply by SayeedChowdhury** [reply]
Hi @jhauret, thanks for sharing your code. Can you point to the implementation you used for the SEANet and streaming SEANet results in your paper? Thanks.

**Reply by jhauret** [reply]
Hi @SayeedChowdhury, since Google has not published their implementation, I made my own:

- Here is [a gist](https://gist.github.com/jhauret/0b12892b8770205d999ac0b1f058156f) for the generator architecture.
- Here is [the code](https://github.com/seungwonpark/melgan/blob/master/model/discriminator.py) I used for the discriminators.
- The training strategy is the same as in EBEN.
 
 
 Note:
The only difference with the papers is that StreamingSeanet uses causal convolutions and I did not. Nevertheless, it is fair to compare with this non-causal version of StreamingSeanet, since it would only decrease the performance.

**Reply by SayeedChowdhury** [reply]
Thanks a lot @jhauret, can  you also refer to the code for the 2 changes you mentioned above?

**Reply by jhauret** [reply]
You're welcome !

If you are referring to the normalized feature loss and the loss balancer, the [code snippets](https://github.com/jhauret/eben/discussions/1#discussioncomment-5384660) are in this conversation thread.

**Reply by SayeedChowdhury** [reply]
Great, thanks a lot!

**Reply by jiaweiru** [reply]
Hi, thanks for contributing the code about SEANet! I was wondering if the downsampling convolution in EncBlock needs to incorporate an activation function, in some other implementations, such as EnCodec, it seems that the activation is performed before the convolution.

```
class EncBlock(nn.Module):
    def __init__(self, out_channels: int, stride: int, weight_norm: bool):
        super().__init__()

        self.residuals = nn.Sequential(
            ResidualUnit(channels=out_channels // 2, dilation=1),
            ResidualUnit(channels=out_channels // 2, dilation=3),
            ResidualUnit(channels=out_channels // 2, dilation=9),
        )
        self.conv = conv1d_wrapper(
            in_channels=out_channels // 2,
            out_channels=out_channels,
            kernel_size=2 * stride,
            stride=stride,
            padding=stride - 1,
            bias=False,
            padding_mode="reflect",
            weight_norm=weight_norm,
        )

    def forward(self, x):
        out = self.conv(self.residuals(x))
        return out
```

**Reply by jhauret** [reply]
Hi @jiaweiru, great question and unfortunately this is the kind of detail we cannot find in the article... I think there is no clear answer, just like for the type of residual unit (pre/post-activation or all other sub-versions). If you have enough computing power, you should try both and keep the best one for your case. If you find an ablation study on this point, don't hesitate to share it!

**Reply by jiaweiru** [reply]
Thanks for the reply, I will try it afterward if I have time.
